File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.kt
import kotlin.math.*

class Solution {    
    var answer: Int = Int.MAX_VALUE
    var n = 0
    lateinit var relations: MutableList<BooleanArray>
    lateinit var visited: BooleanArray

    fun solution(_n: Int, wires: Array<IntArray>): Int {
        n = _n
        relations = MutableList(size = n) { BooleanArray(size  = n) }
        visited = BooleanArray(size = n)
        
        wires.forEach { wire ->
            val (x, y) = wire.map { it - 1 }
            
            relations[x][y] = true
            relations[y][x] = true
        }
    
        
        for ((x, y) in wires) {
            relations[x-1][y-1] = false
            relations[y-1][x-1] = false
            val count = bfs()
            relations[x-1][y-1] = true
            relations[y-1][x-1] = true
            answer = min(answer, abs(n - count*2))
            visited.fill(false)
        }
        
        return answer
    }
    
    
    private fun bfs() : Int{
        val queue = ArrayDeque<Int>()
        queue.add(0)
        visited[0] = true
        var count = 1
        
        while (queue.isNotEmpty()) {
            val num = queue.removeFirst()
            relations[num].forEachIndexed { idx, bol ->
                if (visited[idx].not() && bol) {
                    queue.add(idx)
                    visited[idx] = true
                    count ++
                }
            }
        }
        return count
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
answer = 999999999999

def solution(n, wires):
    relations = [[False for _ in range(n)] for _ in range(n)]
    
    for wire in wires:
        x, y = map(lambda x: x - 1, wire)
        relations[x][y] = True
        relations[y][x] = True

    dfs(n, 0, relations)
    
    return answer


def dfs(n, idx, relations):
    global answer
    count = 1
    
    for i in range(n):
        if relations[idx][i]:
            relations[idx][i] = False
            relations[i][idx] = False
            count += dfs(n, i, relations)
            relations[idx][i] = True
            relations[i][idx] = True
    
    answer = min(answer, abs(n - count - count))

    return count


